Another approach to API

boiler23 opened the pull request:
So, I decided to abandon the idea with multiple WakeLock objects, as there was much negative feedback about it.
Now we consider WakeLock object as a singleton, with acquire/release methods. Each WakeLock would be associated with one type of resource - screen, cpu, wifi, etc., so that you have to acquire `screen.wakeLock`, `system.cpu.wakeLock`, etc. for corresponding resource. Developer can safely acquire it many times and release it only once (or wait for page to unload, or maybe wait for page to exit fullscreen). If he needs reference counting - he should implement it himself. No GC bindings at all :)

Commit:
Example updated to be less confusing

Commit:
Formatting fixed

Commit:
Changed isHeld property to status, added status change event. Added example for using them.

Commit:
New approach with one wake lock for one browsing context.

Commit:
Added example about locks with timeout

Commit:
Corrent reference to Promise spec

Commit:
Renamed lock property to wakeLock

marcoscaceres commented:
There is no `system` object, so that's already a non-starter I'm afraid. What we want is really just:

```
partial interface Navigator {
   promise requestWakeLock(WakeLockType type);
   void releaseWakeLock(WakeLockType type)
}

enum WakeLockType{ "system", "display" }
```

And that's it for now.

marcoscaceres commented:
Actually, I changed my mind... we really want something like this:

```
partial interface Navigator {
   attribute WakeLock wakeLock;
}

interface WakeLock{
   promise request(WakeLockType type);
   promise release(WakeLockType type);
}

enum WakeLockType{ "system", "display" }
```

marcoscaceres commented:
(the above design is much more extensible). It allows us to do a number of interesting things, like add EventHandler if we need to later, plus a way to check if the lock is held by type, etc. in one place. 

However, I still think it's too early to be doing the API design. We need more security people to look a the API because that will fundamentally shape what the API looks like.

mounirlamouri commented:
I like Marcos' proposal. We might need to add an event like `lockchange` and a `WakeLockEvent` that has a `readonly attribute WakeLockType lockType`.

Also, even though `"display"` wake lock doesn't sound appropriate from a Worker, what about `"system"` locks?

boiler23 commented:
Ok. We discussed internally and we agree with Marcos's proposal. However, I don't feel happy about putting everything we don't know where to put into `Navigator`.
Also I'd prefer to name lock types as `"screen"` and `"cpu"`.

Maybe we should also consider requesting many locks by one call, e.g. `navigator.wakeLock.request(['screen', 'cpu'])` or even with timeouts `navigator.wakeLock.request({ screen: 1000, cpu: 5000 })`?

mounirlamouri commented:
Taking an array of WakeLockType in addition of a single WakeLockType would be backward compatible so if the use cases are weak, we could wait or do it in the initial proposal if the use cases are more important.

However, I'm not convinced that we want to set a timeout for the locks. It would be fairly easy to implement in Javascript. It doesn't sound like a so common use cases that we would want to provide this syntax sugar. What did you had in mind?

boiler23 commented:
We discussed it from security point of view.
If something goes wrong or developer forgot to release the lock - it may be a bit more safe to have timeout on a lock.

marcoscaceres commented:
@mounirlamouri I tend to agree with @boiler23 - just because I can see it being a huge mess to self-manage the locks, which would lead to a lot of errors. 

The problem with trying to manage your own locks is that the conventions for delaying the locks on various platforms are different (e.g., mouse move events, touch events, and even orientation-change resets the lock on some platforms - and as a developer you don't know what those conventions are). It basically means you need to set up a whole bunch of different listeners in order to continuously extend the lock for the desired time (this is wasteful and redundant). The UA is in a way better position to handle this. Imagine also a user jumping from one application to another on the phone. The application would only get an `focus` event (which is another you would need to listen for!) and it may be outside the timeout the developer may have originally set. It's really messy.  

@boiler23, if we keep the design simple, it makes it much less likely to lead to interop errors and much more easy to reason about (and to spec!). Having the combined dictionary/timeout thing is nice, but I think it's not necessary. That level of abstraction can be added on-top by a library. The same for having an array - it just adds unnecessary complexity.  

About the name "CPU" - I don't want to start bikeshedding yet, but the reason I personally don't like it is that it might also keep WiFi or other radios on (i.e., it's not just the CPU). It's basically like keeping everything on except the screen.    

@boiler23, about `Navigator`: that's our (the Web's) dumping ground for these kinds of APIs (we've been arguing about this for years and it's just convention now - and it's really not that important). We all hate it - but we don't have anything better until we get JS module support. It's also a predictable place to find this stuff. At least for now, I would say we just keep it there.

marcoscaceres commented:
Ok, I've [sent what we discussed here to the WHATWG](http://lists.whatwg.org/htdig.cgi/whatwg-whatwg.org/2014-July/297231.html).... It's in God's hands now :pray:. I expect the API will probably change fairly radically from a round of WHATWG feedback.

Commit:
Updated spec to match what we discussed in pull request #9

boiler23 commented:
Ok, I've also updated PR to match the results of our discussion.

Commit:
Trivial commit for triggering a gh-pages build

Commit:
Revert "Trivial commit for triggering a gh-pages build"

This reverts commit 1420049b4317317bf533e9e7b952bab4ca6113b4.

marcoscaceres commented:
This needs to do two more rounds of reviews. The TAG wants to take a look and we need to send the API proposal to public-script-coord for an idiomatic check. 

@boiler23 unless you don't mind constantly having to update the PR, I would say hold until we finish the discussions at the WHATWG and W3C on email. We can set a deadline for end of the month (though I am going away on vacation on tuesday this coming week for 2 weeks).

boiler23 commented:
Agreed. Let's set a deadline for the end of month. Have a nice vacation!

@mounirlamouri would it be better to push our current wake locks implementation into Chromium for review, or we should better wait for the end of month (when we'll merge new spec and get it more official)?

marcoscaceres commented:
@boiler23 thanks! And sorry that this feels like we are going through a lot of slow process. It's important to make sure that we have everyone on side and we have strong indications of support from the community.

boiler23 commented:
@marcoscaceres of course, I understand it.

mounirlamouri commented:
@boiler23 it is fine to push your current implementation as part of your Intent to Implement as long as it is clear to you that this implementation is temporary and might likely change in the future.

marcoscaceres commented:
@boiler23 would be also really helpful if you can investigate across platforms if we can do the tiered approach to controlling the wake lock by type. We know already that we can do "display" and "cpu" ... but what about "network"? Is network, as FirefoxOS has it, even a good idea/implementable?

boiler23 commented:
Ok, we will take a look.
On Android it is implementable https://developer.android.com/reference/android/net/wifi/WifiManager.WifiLock.html
I'm not sure it has real use-cases for desktop platforms, but I think there should be a way to do that.

boiler23 commented:
Problem is also actual for Mac OS, it has API for system wake locks:
https://developer.apple.com/library/mac/Documentation/IOKit/Reference/IOPMLib_header_reference/Reference/reference.html#//apple_ref/c/macro/kIOPMAssertionTypePreventSystemSleep
https://developer.apple.com/library/mac/qa/qa1340/_index.html

So, the idea to merge network and cpu into one wake lock looks reasonable for me.

marcoscaceres commented:
Not sure what you mean by the "problem is also actual"? Can you elaborate?

As an aside, I've been trying to get this to work using Swift in OSX, but
haven't figured out how to access IOPMLib sub-module of IOKit :(  need to
ask on stack exchange or something.

It's trivial to do this in iOS, however, as it's just a Boolean you can set
in on the NSApplication.

On Wednesday, July 30, 2014, Ilya Bogdanovich notifications@github.com
wrote:

> Problem is also actual for Mac OS, it has API for system wake locks:
> 
> https://developer.apple.com/library/mac/Documentation/IOKit/Reference/IOPMLib_header_reference/Reference/reference.html#//apple_ref/c/macro/kIOPMAssertionTypePreventSystemSleep
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/w3c/screen-wake/pull/9#issuecomment-50596749.

marcoscaceres commented:
I guess you mean:

> "Assertions are just suggestions to the OS, and the OS can only honor
> them to the best of its ability. In the case of low power or a thermal
> emergency, the system may sleep anyway despite the assertion."

I think that's fine. The thing is just to notify the app that it's lost the
lock through an event.

On Wednesday, July 30, 2014, Marcos Caceres marcos@marcosc.com wrote:

> Not sure what you mean by the "problem is also actual"? Can you elaborate?
> 
> As an aside, I've been trying to get this to work using Swift in OSX, but
> haven't figured out how to access IOPMLib sub-module of IOKit :(  need to
> ask on stack exchange or something.
> 
> It's trivial to do this in iOS, however, as it's just a Boolean you can
> set in on the NSApplication.
> 
> On Wednesday, July 30, 2014, Ilya Bogdanovich <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> 
> > Problem is also actual for Mac OS, it has API for system wake locks:
> > 
> > https://developer.apple.com/library/mac/Documentation/IOKit/Reference/IOPMLib_header_reference/Reference/reference.html#//apple_ref/c/macro/kIOPMAssertionTypePreventSystemSleep
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/w3c/screen-wake/pull/9#issuecomment-50596749.

boiler23 commented:
I was in doubt if there is API for doing that on desktop OS (not screen but system). And Mac OS already  seems to keep both cpu and network by one lock.

boiler23 commented:
So, today is the last day of July, we're reaching the deadline :) Our next steps?

marcoscaceres commented:
So, there are technical and political things for us to take care of.
Politically: we didn't get any negative feedback from the WHATWG... But we
didn't get strong messages of support either (that's ok). On the W3C side I
has a call with the CEO of the W3C to make sure this gets added to the
DAP charter without a whole bunch of pain - looks like that will happen.
Just waiting till Dom is back from vacation. This is a blocker for
publication (who cares), but not a blocker to just do the spec work.

On the technical side, seems that we want to defer "system" for now and
just focus on "screen". It seems we (you, Mounir, and I) are mostly happy
with the API - just need to add the event handler stuff, as Mounir
suggested in the WGATWG thread.

We still need to get our ideas in front of the TAG. I'll email them next
week.

So, from here, we need to write it up and let's get an unofficial draft up.
Ilya, if you start it, I can clean up anything next week. Then Mounir can
give it a good review and sanity check.

On Thursday, July 31, 2014, Ilya Bogdanovich notifications@github.com
wrote:

> So, today is the last day of July, we're reaching the deadline :) Our next
> steps?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/w3c/screen-wake/pull/9#issuecomment-50734658.

mounirlamouri commented:
:+1: with what that guy just said!

boiler23 commented:
:+1: ok, I'll update the draft tomorrow!

Commit:
Using "screen" instead of "system" everywhere

Commit:
Added wakelockchanged event

Commit:
Fixed titles

Commit:
Fixed descriptions for request/release methods. Added information about resolving promises.

Commit:
Fixed descriptions for wake lock types

Commit:
Added isHeld method

Commit:
Added acknowledgments section

boiler23 commented:
So, I made a number of updates to the spec: added an event for lock changes as well as method to get lock status (it also looks necessary). Also I've updated some descriptions. Please take a look!

marcoscaceres commented:
ok, let's go with this as first round. @boiler23, please start using HTML5 Tidy to make sure line lengths are kept at 80 chars (makes it hard to review stuff and to track changes otherwise).

marcoscaceres commented:
I'm also going to squish the history a bit...

boiler23 commented:
ok, got it.
